src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,255 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.graphanalysis;
+
+import com.google.common.base.Predicate;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+/**
+ * Base class for flow scanners, which offers basic methods and stubs for algorithms
+ * Scanners store state internally, and are not thread-safe but are reusable
+ * Scans/analysis of graphs is implemented via internal iteration to allow reusing algorithm bodies
+ * However internal iteration has access to additional information
+ *
+ * Provides 4 sets of common APIs to use, in decreasing expressiveness and increasing genericness:
+ *   - findFirst - find first node match a predicate
+ *   - filteredNodes - return a collection of nodes filtered by a predicate, between heads(inclusive) and blackList (exclusive)
+ *   - visitor - call a visitor on each node we encounter
+ *   - Iterator/filterator based: FlowNode-by-FlowNode walking, after setup with head/blacklist
+ *
+ *  All operations can start with one or more "head" FlowNodes, and walk back from there, stopping when we hit blackList nodes
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public abstract class AbstractFlowScanner implements Iterable <FlowNode>, Filterator<FlowNode> {
+
+    protected FlowNode _current;
+
+    protected FlowNode _next;
+
+    protected Collection<FlowNode> _blackList = Collections.EMPTY_SET;
+
+    /** Convert stop nodes to a collection that can efficiently be checked for membership, handling null if needed */
+    @Nonnull
+    protected Collection<FlowNode> convertToFastCheckable(@CheckForNull Collection<FlowNode> nodeCollection) {
```

Comment:
Syntactically cleaner, yes, but there's a catch: the contains method may be called many, many times (once per FlowNode).  I've seen some posts suggesting the Guava Set/Map implementations are less performant than the JDK ones.  A few examples:

http://stackoverflow.com/questions/12573463/performance-of-guavas-immutableset-contains 

https://groups.google.com/forum/#!topic/guava-discuss/I7yPpa5Hlpg 

I haven't seen a recent comparison, but given this is performance-sensitive and only a few LoC, we should prefer the builtin implementation unless Guava turns out to be faster.